### PR TITLE
Semiautomatic hackage releases

### LINKED ISSUE #1310: Automate hackage releases
* From the recent experience we know that:
  * Users needs a hackage release for every github release we do: #1282
  * The hackage release is cumbersome and error prone and involve the coordination of several contributors, see #1287 and #1175
* So we *have* to automate the process as much as we can, making the automating process itself lightweight if possible

The steps needed to make the release are:

1. Identify changes of each subpackage and check if they can break downstream ones
2. Bump up the package version accordingly and propagate it to the bounds of downstream packages
3. Make pr(o several prs) with those changes, ensuring the order of dependencies
4. Cherry pick those changes onto the github release commit as a branch
5. Check the subpackage builds succesfully in that branch (and test it?)
6. Do a `cabal check && cabal haddock subpackage && cabal sdist subpackage`
7. Do a `cabal upload subpackage` (we can use https://github.com/haskell-actions/hackage-publish)

I think the steps 5, 6 and 7 could be automated in ci, uploading package candidates to do a final review and publish all of them.

Not sure how could we automate 1-4 though. Maybe doing a `git diff` per subpackage and update version bounds of .cabal files with some form of grep&sed (bumping out version as always breaking with no fine grained pvp considerations as suggested by @michaelpj)

There are projects with lot of subpackages and maybe they have been some work to automate this, as pointed by @wz1000.
For example amazonka. Afaics they have a Makefile that includes upload of packages: https://github.com/brendanhay/amazonka/blob/develop/Makefile

### PR DESCRIPTION
**Pull request by jneira**
* This try to make hackage releases as smooth as possible, hopefully fixes #1310
* It is dumb about changes: if some file (module obvious exclusions like docs files) has changed, we have to bomp up the version or it will fail
* It builds all packages for three last ghc major versions so it will take some time (with cold cache)
* It upload package candidates, a human should confirm the definitive release (ahopefully fter some manual checks)
* It depends on #2156 to avoid errors about ghc-api-compat
* ~~It is failing for hls-call-hierarchy-plugin:~~

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2163"></a>

### CODE REVIEW

File: .github/workflows/hackage.yml
Comment (jneira) [inline .github/workflows/hackage.yml:45]: i think we could reuse the test/bench cache here

**Comment by July541** [comment]
It is caused by `aeson`'s `Value`, `Value` support `Ord` instance after their [1.5.2.0](https://hackage.haskell.org/package/aeson-1.5.6.0/docs/src/Data.Aeson.Types.Internal.html#line-398).

I checked your test and found it uses [1.4.7.1](https://github.com/jneira/haskell-language-server/runs/3524694108?check_suite_focus=true#step:11:136)...

I think to add a lower bound for `aeson` in `hls-call-hierarchy-plugin` is enough (Maybe you can add it directly, I don't know how to participate in this thread..)

**Comment by jneira** [comment]
> I think to add a lower bound for `aeson` in `hls-call-hierarchy-plugin` is enough (Maybe you can add it directly, I don't know how to participate in this thread..)

dont worry, the diagnostic is the "hard" part, will add the bound here, thanks!

File: .github/workflows/hackage.yml
Comment (isovector) [inline .github/workflows/hackage.yml:13]: Looks like this list needs to be kept up to date when new plugins are added? Do we have any documentation on how to add new plugins? If so, we should document this fact there too.

Comment (jneira) [reply .github/workflows/hackage.yml:13]: Yeah, fortunately plugins have an unique id and it is relatively easy search in the codebase using it. So search for a similar existing plugin is a required step in front of outdated documentation. But you are right, i guess the best place to list all the places (including .cabal and generic config) could be https://haskell-language-server.readthedocs.io/en/latest/contributing/plugin-tutorial.html#wrapping-up

Comment (jneira) [reply .github/workflows/hackage.yml:13]: @isovector #2166

File: .github/workflows/hackage.yml
Comment (pepeiborra) [reply .github/workflows/hackage.yml:45]: The total cache limit is still 5GB so I would say reusing the test cache would be much preferred to avoid over-running the limit:

https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy

File: .github/workflows/hackage.yml
Comment (pepeiborra) [inline .github/workflows/hackage.yml:95]: These are only needed to build the benchmark suite of ghcide. The benchmark suite is not a requirement for uploading to Hackage imho, and should be skipped

File: .github/workflows/hackage.yml
Comment (pepeiborra) [inline .github/workflows/hackage.yml:101]:
```suggestion
          cabal build --enable-tests
```

File: .github/workflows/hackage.yml
Comment (pepeiborra) [inline .github/workflows/hackage.yml:107]: given that we build them, can we upload the Haddocks too?

**Review by pepeiborra — APPROVED** [review]
Nice job, this is going to be super useful

**Comment by pepeiborra** [comment]
Next step would be integrating @kowainik Policeman to validate the updated version numbers: https://kowainik.github.io/projects/policeman

File: .github/workflows/hackage.yml
Comment (jneira) [reply .github/workflows/hackage.yml:95]: ya, but to use a simple autogenerated cabal.project with `packages: . ../../* ../../plugins/*` cabal has to find a build plan for shake-bench. The alternative is a cabal project listing explicitly all packages but shake-bench.
Otoh in theory you should be able to run the benchmarks (like tests) with the hackage version too, and ideally those allow-newer should be removed to allow it

File: .github/workflows/hackage.yml
Comment (jneira) [reply .github/workflows/hackage.yml:101]: ideally benchmarks should be able to run using the hackage version

**Comment by jneira** [comment]
> 
> 
> Next step would be integrating @kowainik Policeman to validate the updated version numbers: https://kowainik.github.io/projects/policeman

wow, i had forgotten it, it would be really nice

**Comment by jneira** [comment]
I've used a hackage token generated from my account, it can be changed in the project secrets
However i have not perms to upload to all packages involved, missing are:
* [ ] https://hackage.haskell.org/package/shake-bench
* [ ] https://hackage.haskell.org/package/hie-compat
* [ ] https://hackage.haskell.org/package/hls-graph
* [ ] https://hackage.haskell.org/package/hls-ormolu-plugin
* [ ] https://hackage.haskell.org/package/hls-module-name-plugin
* [ ] https://hackage.haskell.org/package/hls-refine-imports-plugin
* [ ] *https://hackage.haskell.org/package/hls-call-hierarchy-plugin*, it is not uploaded

Most of them have @pepeiborra as unique maitainer so an option could be that @pepeiborra  generates a hackage api key (i would create a new, dedicated, one). But even doing that i would add some more maintainers to those plugin with only one parent :wink:

**Comment by pepeiborra** [comment]
I have added you to the maintainers list for all of them

**Comment by jneira** [comment]
Ok the workflow has failed in the last step, cause i did not put the hackage api key:

```
Run haskell-actions/hackage-publish@v1
  with:
    packagesPath: packages
    publish: false
    hackageServer: http://hackage.haskell.org
Uploading hls-module-name-plugin-1.0.0.1 to http://hackage.haskell.org/packages/candidates
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  7777    0     0  100  7777      0  42037 --:--:-- --:--:-- --:--:-- 42037
curl: (22) The requested URL returned error: 401 Unauthorized
Error: Process completed with exit code 22.
```

I think the deinitive test can be done with the next release

**Comment by jneira** [comment]
@MergifyIO backport 1.3.0-hackage

**Comment by pepeiborra** [comment]
It would be nice to back port this to the 1.3.0-hackage branch and use it for the Hackage upload

**Comment by jneira** [comment]
I was waiting for the automatic backport, but CI is quite slow lately, will backport manually

**Comment by jneira** [comment]
backported with 97003e2
ci running in https://github.com/haskell/haskell-language-server/actions/runs/1212661332

**Comment by jneira** [comment]
As the workflow will take ages to start i've started it in my repo: https://github.com/jneira/haskell-language-server/runs/3545669772

**Pull request merged**
